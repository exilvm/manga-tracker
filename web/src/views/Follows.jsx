import React from 'react';
import { lighten, makeStyles } from '@material-ui/core/styles';
import {
  Button,
  Container,
  Grid,
  List,
  ListItem,
  ListItemText,
  Paper,
  Typography,
} from '@material-ui/core';
import NextLink from 'next/link';

import { defaultDateDistanceToNow, followUnfollow } from '../utils/utilities';


const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1,
    padding: theme.spacing(2),
  },
  followCard: {
    height: '100%',
    overflow: 'hidden',
    marginBottom: theme.spacing(2),
    backgroundColor: lighten(theme.palette.background.paper, 0.05),
  },
  followTitle: {
    paddingTop: theme.spacing(2),
    paddingLeft: theme.spacing(2),
  },
  followContent: {
    display: 'flex',
  },
  followDetails: {
    display: 'flex',
    flexDirection: 'column',
    paddingLeft: theme.spacing(2),
  },
  thumbnail: {
    paddingLeft: theme.spacing(2),
    maxWidth: '200px',
    height: '250px',
    [theme.breakpoints.down('sm')]: {
      maxWidth: '180px',
    },
    [theme.breakpoints.down('xs')]: {
      maxWidth: '145px',
      height: '175px',
    },
  },
  serviceList: {
    overflow: 'auto',
    maxHeight: '250px',
  },
  followService: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  serviceName: {
    marginRight: theme.spacing(2),
  },
}));

function Follows(props) {
  const {
    follows = [],
  } = props;

  const classes = useStyles();
  const columnsXs = 1;
  const columnsMd = 2;

  const renderFollow = (follow) => {
    const followedServices = follow.followed_services;

    return (
      <Grid item xs={12/columnsXs} md={12/columnsMd} key={follow.manga_id}>
        <Paper className={classes.followCard}>
          <Typography className={classes.followTitle}>{follow.title}</Typography>
          <div className={classes.followContent}>
            <NextLink href='/manga/[id]' as={`/manga/${follow.manga_id}`} passHref>
              <a target='_blank'>
                <img
                  src={follow.cover}
                  className={classes.thumbnail}
                  alt={follow.title}
                />
              </a>
            </NextLink>
            <div className={classes.followDetails}>
              <table>
                <tbody>
                  <tr>
                    <td>
                      <Typography>Latest release: </Typography>
                    </td>
                    <td>
                      <Typography>
                        {defaultDateDistanceToNow(new Date(follow.latest_release))}
                      </Typography>
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <Typography>Latest chapter: </Typography>
                    </td>
                    <td>
                      <Typography>{follow.latest_chapter || 'No chapters'}</Typography>
                    </td>
                  </tr>
                </tbody>
              </table>
              <List className={classes.serviceList}>
                <ListItem key='all_services' className={classes.followService} disableGutters>
                  <ListItemText primary='All services' className={classes.serviceName} />
                  <Button variant='contained' color='primary' onClick={followUnfollow(follow.manga_id)}>
                    {followedServices.indexOf(null) < 0 ? 'Follow' : 'Unfollow'}
                  </Button>
                </ListItem>
                {follow.services.map((service) => (
                  <ListItem key={service.service_id} className={classes.followService} disableGutters>
                    <ListItemText primary={service.service_name} className={classes.serviceName} />
                    <Button variant='contained' color='primary' onClick={followUnfollow(follow.manga_id, service.service_id)}>
                      {followedServices.indexOf(service.service_id) < 0 ? 'Follow' : 'Unfollow'}
                    </Button>
                  </ListItem>
                ))}
              </List>
            </div>
          </div>
        </Paper>
      </Grid>
    );
  };

  return (

    <Container maxWidth='lg' disableGutters>
      <Paper className={classes.root}>
        <Grid container spacing={1}>
          {follows.map(renderFollow)}
        </Grid>
      </Paper>
    </Container>
  );
}

export default Follows;
