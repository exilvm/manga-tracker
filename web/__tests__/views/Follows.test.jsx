import React from 'react';
import renderer from 'react-test-renderer';
import Follows from '../../src/views/Follows';

describe('Follows page should render correctly', () => {
  const follows = [
    {
      title: 'Dr. STONE',
      cover: 'https://mangadex.org/images/manga/20882.jpg?1585634146',
      manga_id: 1,
      services: [
        {
          service_id: 1,
          service_name: 'MANGA Plus',
          title_id: '100010',
          url: 'https://mangaplus.shueisha.co.jp/titles/{}',
        },
        {
          service_id: 2,
          service_name: 'MangaDex',
          title_id: '20882',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followed_services: [
        1,
        null,
      ],
      },
    {
      title: 'Shuumatsu no Valkyrie',
      cover: 'https://mangadex.org/images/manga/33537.jpg?1591183509',
      manga_id: 8,
      services: [
        {
          service_id: 2,
          service_name: 'MangaDex',
          title_id: '33537',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followed_services: [
        null,
      ],
    },
    {
      title: 'Kengan Omega',
      cover: 'https://mangadex.org/images/manga/33538.jpg?1593474917',
      manga_id: 64,
      services: [
        {
          service_id: 2,
          service_name: 'MangaDex',
          title_id: '33538',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followed_services: [
        null,
      ],
    },
  ];

  it('should render correctly without follows', () => {
    const tree = renderer
      .create(<Follows />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with follows', () => {
    const tree = renderer
      .create(<Follows follows={follows} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
