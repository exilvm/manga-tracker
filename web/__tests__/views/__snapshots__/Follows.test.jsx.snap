// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Follows page should render correctly should render correctly with follows 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <div
          className="MuiPaper-root makeStyles-followCard-2 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-followTitle-3 MuiTypography-body1"
          >
            Dr. STONE
          </p>
          <div
            className="makeStyles-followContent-4"
          >
            <a
              href="/manga/1"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <img
                alt="Dr. STONE"
                className="makeStyles-thumbnail-6"
                src="https://mangadex.org/images/manga/20882.jpg?1585634146"
              />
            </a>
            <div
              className="makeStyles-followDetails-5"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest release: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        1 day ago
                      </p>
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest chapter: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        90
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
              <ul
                className="MuiList-root makeStyles-serviceList-7 MuiList-padding"
              >
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      All services
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Unfollow
                    </span>
                  </button>
                </li>
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      MANGA Plus
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Unfollow
                    </span>
                  </button>
                </li>
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      MangaDex
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Follow
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <div
          className="MuiPaper-root makeStyles-followCard-2 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-followTitle-3 MuiTypography-body1"
          >
            Shuumatsu no Valkyrie
          </p>
          <div
            className="makeStyles-followContent-4"
          >
            <a
              href="/manga/8"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <img
                alt="Shuumatsu no Valkyrie"
                className="makeStyles-thumbnail-6"
                src="https://mangadex.org/images/manga/33537.jpg?1591183509"
              />
            </a>
            <div
              className="makeStyles-followDetails-5"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest release: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Unknown
                      </p>
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest chapter: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        No chapters
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
              <ul
                className="MuiList-root makeStyles-serviceList-7 MuiList-padding"
              >
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      All services
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Unfollow
                    </span>
                  </button>
                </li>
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      MangaDex
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Follow
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <div
          className="MuiPaper-root makeStyles-followCard-2 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-followTitle-3 MuiTypography-body1"
          >
            Kengan Omega
          </p>
          <div
            className="makeStyles-followContent-4"
          >
            <a
              href="/manga/64"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <img
                alt="Kengan Omega"
                className="makeStyles-thumbnail-6"
                src="https://mangadex.org/images/manga/33538.jpg?1593474917"
              />
            </a>
            <div
              className="makeStyles-followDetails-5"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest release: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Unknown
                      </p>
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        Latest chapter: 
                      </p>
                    </td>
                    <td>
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        10
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
              <ul
                className="MuiList-root makeStyles-serviceList-7 MuiList-padding"
              >
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      All services
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Unfollow
                    </span>
                  </button>
                </li>
                <li
                  className="MuiListItem-root makeStyles-followService-8"
                  disabled={false}
                >
                  <div
                    className="MuiListItemText-root makeStyles-serviceName-9"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      MangaDex
                    </span>
                  </div>
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Follow
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Follows page should render correctly should render correctly without follows 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
    />
  </div>
</div>
`;
